core-v1/src/main/scala/dsl/JustifiedBuildExprDsl.scala

```diff
@@ -0,0 +1,119 @@
+package com.rallyhealth.vapors.v1
+
+package dsl
+
+import algebra._
+import data.{FactTypeSet, Justified}
+import logic.Negation
+
+import cats.data.NonEmptyList
+import cats.{Foldable, Functor}
+
+trait JustifiedBuildExprDsl extends WrappedBuildExprDsl with JustifiedDslTypes {
+
+  override protected implicit final def compareWrapped: CompareWrapped[Justified] = CompareWrapped.justified
+
+  override protected implicit final def windowComparable: WindowComparable[Justified, OP] = WindowComparable.justified
+
+  override protected implicit final def extract: Extract[Justified] = Extract.justified
+
+  override protected implicit final def wrapConst: WrapConst[Justified] = WrapConst.justified
+
+  override protected def wrapElement[C[_], A](
+    outer: Justified[C[A]],
+    element: A,
+  ): Justified[A] =
+    Justified.byInference("elementOf", element, NonEmptyList.of(outer))
+
+  // TODO: Should this be visible outside this trait?
+  protected def dontShortCircuit: Boolean = false
+
+  override def not[I, O : OPW](
+    expr: Justified[I] ~:> Justified[O],
+  )(implicit
+    negation: Negation[Justified[O]],
+  ): Expr.Not[Justified[I], Justified[O], OP] =
+    Expr.Not(expr)
+
+  override def valuesOfType[T](
+    factTypeSet: FactTypeSet[T],
+  )(implicit
+    opTs: OP[Seq[Justified[T]]],
+  ): Expr.ValuesOfType[T, Justified[T], OP] =
+    Expr.ValuesOfType[T, Justified[T], OP](factTypeSet, Justified.ByFact(_))
+
+  override type SpecificHkExprBuilder[I, C[_], A] = JustifiedHkExprBuilder[I, C, A]
+
+  override implicit def hk[I, C[_], A](expr: I ~:> C[Justified[A]]): JustifiedHkExprBuilder[I, C, A] =
+    new JustifiedHkExprBuilder(expr)
+
+  final class JustifiedHkExprBuilder[I, C[_], A](override protected val inputExpr: I ~:> C[Justified[A]])
+    extends HkExprBuilder[I, C, A] {
+
+    override def exists(
+      conditionExprBuilder: Justified[A] =~:> Justified[Boolean],
+    )(implicit
+      opO: OP[C[Justified[A]]],
+      opA: OP[Justified[A]],
+      opB: OP[Justified[Boolean]],
+      foldC: Foldable[C],
+    ): Ap[I, C[Justified[A]], Justified[Boolean]] =
+      Expr.AndThen(
+        inputExpr,
+        Expr.Exists[C, Justified[A], Justified[Boolean], OP](
+          conditionExprBuilder(Expr.Identity()),
+          combineTrue = Justified.byInference("exists", true, _),
+          combineFalse = NonEmptyList
+            .fromList(_)
+            .map { justified =>
+              Justified.byInference("exists", false, justified)
+            }
+            .getOrElse {
+              // exists in an empty collection is false
+              // TODO: Should I put a reason instead of just a const?
+              //       Maybe I should pass the original value in these functions?
+              Justified.byConst(false)
```

Comment:
I think `Justified.byConst` should probably take a reason string. It can default to where the const is defined in code or a constant string like `"source code"` or something

Reply:
This is probably worth doing before merging.